packages/runtime/container-runtime/src/gc/gcTelemetryTracker.ts

```diff
@@ -0,0 +1,353 @@
+		private readonly getNodePackagePath: (
+			nodePath: string,
+		) => Promise<readonly string[] | undefined>,
+	) {}
+
+	/**
+	 * Returns whether an event should be logged for a not that isn't active anymore. Some scenarios where we won't log:
+	 * 1. When a DDS is changed or loaded. The corresponding data store's event will be logged instead.
+	 * 2. An event is logged only once per container instance per event per node.
+	 */
+	private shouldLogNonActiveEvent(
+		nodeId: string,
+		nodeType: GCNodeType,
+		usageType: NodeUsageType,
+		nodeStateTracker: UnreferencedStateTracker,
+	) {
+		if (nodeStateTracker.state === UnreferencedState.Active) {
+			return false;
+		}
+
+		// For sub data store (DDS) nodes, if they are changed or loaded, its data store will also be changed or loaded,
+		// so skip logging to make the telemetry less noisy.
+		if (nodeType === GCNodeType.SubDataStore && usageType !== "Revived") {
+			return false;
+		}
+		if (nodeType === GCNodeType.Other) {
+			return false;
+		}
+
+		const uniqueEventId = `${nodeStateTracker.state}-${nodeId}-${usageType}`;
+		if (this.loggedUnreferencedEvents.has(uniqueEventId)) {
+			return false;
+		}
+		this.loggedUnreferencedEvents.add(uniqueEventId);
+		return true;
+	}
+
+	/**
+	 * Called when a node is used. If the node is not active, log an event indicating object is used when its not active.
+	 */
+	public nodeUsed(nodeUsageProps: INodeUsageProps) {
+		// If there is no reference timestamp to work with, no ops have been processed after creation. If so, skip
+		// logging as nothing interesting would have happened worth logging.
+		// If the node is not unreferenced, skip logging.
+		const nodeStateTracker = this.getNodeStateTracker(nodeUsageProps.nodeId);
+		if (!nodeStateTracker || nodeUsageProps.currentReferenceTimestampMs === undefined) {
+			return;
+		}
+
+		const nodeType = this.getNodeType(nodeUsageProps.nodeId);
+		if (
+			!this.shouldLogNonActiveEvent(
+				nodeUsageProps.nodeId,
+				nodeType,
+				nodeUsageProps.usageType,
+				nodeStateTracker,
+			)
+		) {
+			return;
+		}
+
+		const state = nodeStateTracker.state;
+		const { usageType, currentReferenceTimestampMs, packagePath, ...propsToLog } =
+			nodeUsageProps;
+		const eventProps: Omit<IUnreferencedEventProps, "state" | "usageType"> = {
+			id: nodeUsageProps.nodeId,
+			type: nodeType,
+			unrefTime: nodeStateTracker.unreferencedTimestampMs,
+			age:
+				nodeUsageProps.currentReferenceTimestampMs -
+				nodeStateTracker.unreferencedTimestampMs,
+			timeout:
+				state === UnreferencedState.Inactive
+					? this.configs.inactiveTimeoutMs
+					: this.configs.sweepTimeoutMs,
+			...propsToLog,
+			...this.createContainerMetadata,
+		};
+
+		// This will log the following events:
+		// GC_Tombstone_DataStore_Revived, GC_Tombstone_SubDataStore_Revived, GC_Tombstone_Blob_Revived
+		if (nodeUsageProps.usageType === "Revived" && nodeUsageProps.isTombstoned) {
+			sendGCUnexpectedUsageEvent(
```

Comment:
Wanna move this helper to this file? Not sure, just a thought. If so probably rename the file to gcTelemetry.ts.

Reply:
yeah, sounds like a good idea.